api/users/views.py

```diff
@@ -655,3 +658,96 @@ def post(self, request, *args, **kwargs):
             raise ValidationError('Must either be logged in or specify claim email.')
 
         return Response(status=status.HTTP_204_NO_CONTENT)
+
+
+class UserEmailsList(JSONAPIBaseView, generics.ListAPIView, generics.CreateAPIView, UserMixin):
+    permission_classes = (
+        drf_permissions.IsAuthenticatedOrReadOnly,
+        base_permissions.TokenHasScope,
+        CurrentUser,
+    )
+
+    required_read_scopes = [CoreScopes.USER_SETTINGS_READ]
+    required_write_scopes = [CoreScopes.USER_SETTINGS_WRITE]
+
+    view_category = 'users'
+    view_name = 'user-emails'
+
+    serializer_class = UserEmailsSerializer
+
+    # overrides ListAPIViewa
+    def get_queryset(self):
+        user = self.get_user()
+        serialized_emails = []
+        for email in user.emails.all():
+            primary = email.address == user.username
+            hashed_id = hashids.encode(email.id)
+            serialized_email = UserEmail(email_id=hashed_id, address=email.address, confirmed=True, primary=primary)
+            serialized_emails.append(serialized_email)
+        email_verifications = user.email_verifications or []
+        for token in email_verifications:
+            detail = user.email_verifications[token]
+            serialized_unconfirmed_email = UserEmail(email_id=token, address=detail['email'], confirmed=detail['confirmed'], primary=False)
+            serialized_emails.append(serialized_unconfirmed_email)
+
+        return serialized_emails
+
+
+class UserEmailsDetail(JSONAPIBaseView, generics.RetrieveUpdateDestroyAPIView, UserMixin):
+    permission_classes = (
+        drf_permissions.IsAuthenticatedOrReadOnly,
+        base_permissions.TokenHasScope,
+        CurrentUser,
+    )
+
+    required_read_scopes = [CoreScopes.USER_SETTINGS_READ]
+    required_write_scopes = [CoreScopes.USER_SETTINGS_WRITE]
+
+    view_category = 'users'
+    view_name = 'user-email-detail'
+
+    serializer_class = UserEmailsSerializer
+
+    # Overrides RetrieveUpdateDestroyAPIView
+    def get_object(self):
+        email_id = self.kwargs['email_id']
+        user = self.get_user()
+
+        # check to see if it's a confirmed email with hashed id
+        decoded_id = hashids.decode(email_id)
+        if decoded_id:
+            try:
```

Comment:
Nit: Ideally, `try` blocks should only contain the code that is expected to raise an error.

```python
try:
    email = user.emails.get(id=decoded_id[0])
except Email.DoesNotExist:
    email = None
else:
    primary = email.address == user.username
    address = email.address
    confirmed = True

```